pkg/arvo/app/chat-cli.hoon

```diff
@@ -0,0 +1,1111 @@
+  ::    if input is invalid, +slug is called.
+  ::    otherwise, the appropriate work is done and
+  ::    the command (if any) gets echoed to the user.
+  ::
+  ++  obey
+    ^-  (quip move _this)
+    =+  buf=buf.state.cli
+    =+  fix=(sanity [%nop ~] buf)
+    ?^  lit.fix
+      (slug fix)
+    =+  jub=(rust (tufa buf) read)
+    ?~  jub  [[(effect:sh-out %bel ~) ~] this]
+    =^  cal  state.cli  (~(transmit sole-lib state.cli) [%set ~])
+    =^  moves  this  (work u.jub)
+    :_  this
+    %+  weld
+      ^-  (list move)
+      ::  echo commands into scrollback
+      ?.  =(`0 (find ";" buf))  ~
+      [(note:sh-out (tufa `(list @)`buf)) ~]
+    :_  moves
+    %+  effect:sh-out  %mor
+    :~  [%nex ~]
+        [%det cal]
+    ==
+  ::  +work: run user command
+  ::
+  ++  work
+    |=  job=command
+    ^-  (quip move _this)
+    |^  ?-  -.job
+          %target    (set-target +.job)
+          %say       (say +.job)
+          %eval      (eval +.job)
+        ::
+          %join      (join +.job)
+          %leave     (leave +.job)
+          %create    (create +.job)
+        ::
+          %bind      (bind-glyph +.job)
+          %unbind    (unbind-glyph +.job)
+          %what      (lookup-glyph +.job)
+        ::
+          %settings  show-settings
+          %set       (set-setting +.job)
+          %unset     (unset-setting +.job)
+          %width     (set-width +.job)
+          %timezone  (set-timezone +.job)
+        ::
+          %select    (select +.job)
+          %chats     chats
+          %help      help
+        ==
+    ::  +act: build action move
+    ::
+    ++  act
+      |=  [what=term app=term =out-action]
+      ^-  move
+      :*  ost.bowl
+          %poke
+          /cli-command/[what]
+          [our-self app]
+          out-action
+      ==
+    ::  +set-target: set audience, update prompt
+    ::
+    ++  set-target
+      |=  tars=(set target)
+      ^-  (quip move _this)
+      =.  audience  tars
+      [[prompt:sh-out ~] this]
+    ::  +create: new local mailbox
+    ::
+    ++  create
+      ::TODO  configurable security
+      |=  [=path gyf=(unit char)]
+      ^-  (quip move _this)
+      ::TODO  check if already exists
+      =/  =target  [our-self path]
+      =^  moz  this
+        ?.  ?=(^ gyf)  [~ this]
+        (bind-glyph u.gyf target)
+      =-  [[- moz] this(audience [target ~ ~])]
+      %^  act  %do-create  %chat-view
+      :-  %chat-view-action
+      [%create path %channel ~ ~]
+    ::  +join: sync with remote mailbox
+    ::
+    ++  join
+      |=  [=target gyf=(unit char)]
+      ^-  (quip move _this)
+      =^  moz  this
+        ?.  ?=(^ gyf)  [~ this]
+        (bind-glyph u.gyf target)
+      =-  [[- moz] this(audience [target ~ ~])]
+      %^  act  %do-join  %chat-hook
+      :-  %chat-hook-action
+      [%add-synced target]
+    ::  +leave: unsync & destroy mailbox
+    ::
+    ::TODO  but if we leave our own circle, then it disappears for everyone?
```

Comment:
Good point. I'd suggest not exposing %leave from our own circle as a command, and only allowing %leave for foreign circles.

Reply:
Not being able to "silence" local circles seems like a problem to me. Perhaps the solution here is to do per-mailbox subscriptions instead of to `/all`, but that seems maybe outside the scope of this PR.

Reply:
What do you mean "silence"? Like, not render their messages in %chat-cli? Seems like a feature/setting to be stored within %chat-cli.

Reply:
Doesn't seem necessary to switch to a per-mailbox subscription to implement, either.